bgpd/bgp_fsm.c

```diff
@@ -180,6 +180,15 @@ static struct peer *peer_xfer_conn(struct peer *from_peer)
 	 * otherwise we'll get spurious failures during session establishment.
 	 */
 	frr_with_mutex(&peer->io_mtx, &from_peer->io_mtx) {
+		/*
+		 * Before exchanging FD remove doppelgnger from
+		 * keepalive peer hash, if config node peer fd
+		 * is set to -1. If blocked on lock then keepalive
+		 * thread can access doppelgnger pointer with fd -1.
+		 */
+		if(peer->fd < 0)
```

Comment:
`if(` needs a space.

Reply:
If I understood your explanation correctly, would it be correct to simply turn off keepalives (unconditionally) on line 165? I believe that's what we want here.

Reply:
> If I understood your explanation correctly, would it be correct to simply turn off keepalives (unconditionally) on line 165? I believe that's what we want here.

Thanks for review. In bgp_establish we would eventually call bgp_stop for doppelganger and hence it will get removed. It is required to be removed now itself if FD was already set to -1. Like you said it will not cause any harm if we remove it  unconditionally. I will make the changes for the same.

Reply:
@qlyoung I have taken care of you review comments. Can you please review it again?